### ISSUE TITLE
search results in a tree view

### ISSUE THREAD

**Issue by ThomasJuhasz** [Microsoft/vscode#20224]
Hi guys,

I'd really like to have an option to see the search results in a tree view. When working on big JS codebases I find it difficult to see the relevant findings in this flattened list, although in many cases when the search result is small, this flattened list is easier for the eye. 
BTW this search panel could be a nice extension point as well, not sure what is an easier thing to do.
I'll try to contribute sometime, but I hope someone would take this ticket soon :)

Thanks guys,
Thomas

PS.: VSCode is awesome, you are awesome, thank you.

**Comment by dakaraphi** [comment]
I think there are many different interesting ways the results could be represented that are useful to different people based on context of their project.  I also struggle a bit on large codebases.
I recently posted some proposals in #19939 
Possibly some of these ideas could be consolidated, or an extension point made so that alternative views could be provided.

**Comment by Tyriar** [comment]
I think filtering the explorer would cover this https://github.com/Microsoft/vscode/issues/10026

/cc @roblourens since it's related to search though.

**Comment by ThomasJuhasz** [comment]
@Tyriar I really dont think the two are the same. The feature you are suggesting is in visual studio and I cant describe how bad I hated it and it wasn't just me. It's nice to have your explorer untouched while searching in another tab. I think the duplicate tag should be taken off, it's not the same.

**Comment by Tyriar** [comment]
Ping @roblourens

**Comment by jsynowiec** [comment]
@Tyriar anything new regarding this FR? Doing an extensive refactor over a large codebase is really inconvenient without an ability to group results by folder, display a tree or any other additional way of further narrowing down search results.

**Comment by roblourens** [comment]
We are thinking about how search results will look in a multi-folder situation and how that will affect the UX, so we probably won't do anything here until then.

**Comment by seanpoulter** [comment]
Let me know if I can help @roblourens.

**Comment by davidwindell** [comment]
Same here, just came to see if there was a way to match Netbeans' behaviour below that allows toggling a file tree:

![deacf906-414f-11e7-9b4f-e880fdece5b0](https://user-images.githubusercontent.com/1720090/45753430-c0815480-bc10-11e8-8c41-cf705feac4d9.png)

**Comment by seanpoulter** [comment]
I'll put it out there again -- if this something we should get going in a PR @roblourens? I've been using the editor pretty much every day for the last year and it's the only one thing I'd change in the core.

**Comment by zenyuk** [comment]
+1

**Comment by iurimattos** [comment]
+1

**Comment by aaltaras** [comment]
any update or progress on that issue?

**Comment by Sibitu89** [comment]
any update or progress on that issue? This is very usefull feature. When you are looking for a sentense and you know that it should be somewhere in one directory you can easily find it using tree view.

**Comment by byyo** [comment]
+1

**Comment by imaimai86** [comment]
+1

**Comment by sebastienpa** [comment]
+1

**Comment by szepeu** [comment]
Hey, this is really annoying that VS Code, being so great overall, still lacks a feature that is obvious in other editors...

**Comment by ZiTAL** [comment]
It will be good not for search only, to svn/git it will be great too

**Comment by NorbertNemec** [comment]
+1

**Comment by MichaelRo** [comment]
+1

**Comment by iamandrewluca** [comment]
@ZiTAL for Git this is supported from [v1.39.1](https://github.com/microsoft/vscode/releases/tag/1.39.1)
Maybe soon search will also support this.

**Comment by str** [comment]
I was really exited to see the tree widget in the source control/git view. Please include the same feature in the search results view

![image](https://user-images.githubusercontent.com/123817/67488371-f4595580-f62c-11e9-942f-b7d9a2f7bf62.png)

**Comment by chaanto** [comment]
Any update for search result in tree view ?

**Comment by yeegor** [comment]
I will just write something here so that the VSCode team notices us :)

**Comment by MathieuDerelle** [comment]
any progress on this ?

**Comment by NorbertNemec** [comment]
please!!!

**Comment by raitucarp** [comment]
I'm really sorry to bump this issue,  I hope VSCode team implements this feature in the near future without extensions

**Comment by austinsmorris** [comment]
Pretty please?

**Comment by Mirzero** [comment]
Like many others, I would find this feature immensely useful.

**Comment by sibitu** [comment]
Tell my what I can use and I will try. Lets do it together.

czw., 3 cze 2021, 19:56 użytkownik Mirzero ***@***.***>
napisał:

**Comment by cforce** [comment]
Seems like - its easier to change the editor

**Comment by Moon-Raven** [comment]
I would also appreaciate this feature very much!

**Comment by Rory-Powell** [comment]
This would be amazing!

**Comment by matusferko** [comment]
+1 btw this request is only 4 years old 👎

**Comment by ravidbro** [comment]
+1

**Comment by mcintyret** [comment]
Guess I'll keep switching to IntelliJ when I need this feature 😞

**Comment by Twoleeland** [comment]
You should improve the Search functionality despite the low upvotes for this feature and other similar requests.
I use Visual Studio Code often for searching all our projects for particular strings.
Doing the same in VS 2019 is awfully slow compared to VS Code:
starting VS2019, loading all the extensions, waiting till it's ready, opening the "workspace"-folder, then use "find & replace", etc.
VS Code is a lot faster, but the missing grouping options are REALLY counter-productive...

And the time-depended upvoting for feature requests is also counter-productive. Please, let it open, so people can vote anytime, and don't close it because time runs out... Close it only if it's implemented or not possible or replaced by another better feature... 😉

Because in reality:
How many devs do you think have the time to watch every now and then if any software they're using has a feature-request they could upvote for?
You can make a test and count for yourself how many dev-software you are using (including plugins, extensions, not directly dev-realted software, like browsers, and so on) which have a similar feature-request-upvote-systems. And then, calculate, how long you would need to find all the upvote-systems, read through all feature-requests for each upvote-system, and vote for the one or the other feature (after reading through), and so on. Result: no dev on the world has so much time in life left to do this!!! 😅

**Comment by 1ohnny** [comment]
+1

**Comment by vosamoilenko** [comment]
+1

**Comment by chromey** [comment]
I know people say that all the time about their favourite feature request :) ... but it's baffling to me that this is still not implemented. With all the focus on productivity that VS Code usually exercises, this seems like such a no-brainer. Please make it happen.

**Comment by elseifthen** [comment]
This is really the only thing missing that is keeping me from switching from Eclipse to vscode for good as the presentation of project search results is still far superior in Eclipse (I'm still using a >10 year old version).

I do have to search quite often in my large projects and it would make spotting the relevant search results so much easier if they were grouped hierarchically according to the folder structure. If this could be made configurable, everyone could be made happy (o;
I'd attempt to implement it myself but I have no idea how/where to start (I checked src/vs/workbench/contrib/search/browser/searchView.ts but I don't think that's the right place…).

**Comment by davidgulyas** [comment]
I recently switched from VS to VS Code. Not having the search results displayed in a folder tree and not being able to collapse all results in, for example, the test folder, is a huge decrease in usability when exploring a new large code base.

**Comment by cheynove** [comment]
On searches with 50+ results and not having collapse all and a tree view makes it so hard :( please have this implemented soon

**Comment by zivmarks123** [comment]
> On searches with 50+ results and not having collapse all and a tree view makes it so hard :( please have this implemented soon

you do have a "collapse all" button
![image](https://user-images.githubusercontent.com/26737923/163197625-d47b781d-a700-4f25-bd45-3149ce312a05.png)

but the treeview is a must

**Comment by leo-diehl** [comment]
Any updates on this? I kept searching through the settings because I couldn't believe that such a no brainer like this, with clear benefits, hasn't been implemented yet

**Comment by owengibbins** [comment]
Wow, just getting going with VS Code after being a bit dismayed by XCode, and had high hopes until realizing that search results are unnavigable—and seeing that this issue, with so much interest in it, has sat ignored for half a decade is not renewing those hopes at all….

**Comment by 0Grit** [comment]
I need this.

**Comment by seanpoulter** [comment]
@roblourens and the Code team: are we still blocked by a UX decision or is this something we can contribute to? The multi-root workspaces has been implemented - https://github.com/microsoft/vscode/issues/396#issuecomment-341387730 - and so has the Toggle View Mode in VCS mentioned above - https://github.com/microsoft/vscode/issues/20224#issuecomment-545909199. If we get some clarity I expect the folks on the thread can help contribute.

**Comment by ctsstc** [comment]
It seems like the same UI that's used for the Source Control panel could be utilized here, and that the functionality already exists/has been determined if you want to keep it consistent. 

<img width="149" alt="image" src="https://user-images.githubusercontent.com/416814/171943684-6519ae4f-4851-475e-b452-487ddb404215.png">

<img width="194" alt="image" src="https://user-images.githubusercontent.com/416814/171943722-33dc1bb7-359e-4161-9ef1-cdf69e8109ce.png">

**Comment by yeegor-restrata** [comment]
@roblourens, since this feature has been getting requested consistently since 2017 (5 years ago), it'd be perfect if VSCode team could greenlight a contribution on this with an answer to @seanpoulter's question. VSCode is awesome, but as the codebase grows bigger it indeed gets very hard to navigate. Tree view would majorly simplify the process.

**Comment by ryanwinter** [comment]
Ug, I hate parsing through the search list and having to look at the weird path metadata rather than just show me a tree structure.

**Comment by matanhavi** [comment]
Here is another thumb up for this feature👍
If I only know how to write extension for vsCode...

**Comment by grumBit** [comment]
I would love this feature please. 

I compensate for the lack of it using egrep wrapped in a shell script with loads of exclusions, so at least the file path appears first, which makes it somewhat easier to get a sense of the directory tree. Although, it seems a bit poor when the command line is more helpful than vs code!

**Comment by welcometochristown** [comment]
I would also like this feature, web you have different projects in one repository, it can get annoying having to keep changing which files to include. A tree structure would help alot.

**Comment by jakemschiller** [comment]
+1 for this feature

**Comment by carl-lindquist** [comment]
Big upvote on this

**Comment by pdeng6** [comment]
+1 for this feature

**Comment by fm-swe** [comment]
holy shit, how can this still be missing after 5 years... back to eclipse it is

**Comment by trekhleb** [comment]
I've played around trying to implement a **Search Tree** extension here [trekhleb/vscode-search-tree](https://github.com/trekhleb/vscode-search-tree).

However, there are some challenges like:

- not being able to re-use the VSCode search API (I don't quite like the idea of introducing a custom search engine),
- extending the existing Search View instead of implementing a new one, 
- etc.

So the code of the extension is pretty raw, not tested, not optimized, and not ready for publishing yet. But here is a quick demo of what I've ended up with so far:

![](https://raw.githubusercontent.com/trekhleb/vscode-search-tree/main/assets/demo-01.gif)

I'm pausing the work on the extension for now, since the current issue has been picked up for consideration in https://github.com/microsoft/vscode/issues/157454. Hopefully, the Tree View will be implemented in the VSCode core. But meanwhile, feel free to check the source code I've got and extend it if you want.

**Comment by grumBit** [comment]
Fantastic @andreamah ! Thank you for adding that in #161022 .

**Comment by sibitu** [comment]
Thank you :D
